### PR TITLE
refactor: rework `JvmMethod`

### PR DESCRIPTION
**Pull request by magnus-madsen**
This splits `JvmMethod` into two: `JvmMethodSpec` and `JvmMethodImpl`. 

Now `NewObject` depends on `JvmMethodImpl`, whereas `anonClasses` uses `JvmMethodSpec`.

This is in preparation for removal of the `ErasedAst` phase.

### CODE REVIEW

**Comment by magnus-madsen** [comment]
@mlutze CC for speedy approval :)

File: main/src/ca/uwaterloo/flix/language/phase/Eraser.scala [hunk 1]
```diff
@@ -19,38 +19,40 @@ package ca.uwaterloo.flix.language.phase
...
-    //
-    // A mutable set to hold all type information about all closures in the AST.
-    //
-    implicit val ctx: Context = Context(mutable.Set.empty)
-
     val defs = root.defs.map { case (k, v) => k -> visitDef(v) }
     val enums = root.enums.map { case (k, v) => k -> visitEnum(v) }
 
-    ErasedAst.Root(defs, enums, root.entryPoint, root.sources, ctx.anonClasses.toSet)
+    val anonClasses = root.anonClasses.map {
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/Eraser.scala:30]: Would be prettier with a helper like with the `defs` and `enums`, but OK.

File: main/src/ca/uwaterloo/flix/language/phase/Reducer.scala [hunk 2]
```diff
@@ -144,6 +148,14 @@ object Reducer {
     case LiftedAst.Expr.NewObject(name, clazz, tpe, purity, methods, loc) =>
       val ms = methods.map(visitJvmMethod)
       val t = visitType(tpe)
+      val sigs = methods.map {
+        case LiftedAst.JvmMethod(ident, fparams, _, retTpe, purity, loc) =>
+          val f = fparams.map(visitFormalParam)
+          val rt = visitType(retTpe)
+          ReducedAst.JvmMethodSpec(ident, f, rt, purity, loc)
+      }
+      ctx.anonClasses += ReducedAst.AnonClass(name, clazz, t, sigs, loc)
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/Reducer.scala:157]: Again, `hashCode` needed IMO

**Review by mlutze — APPROVED** [review]
I'm OK with this merging now, since it works and doesn't regress from what we had. But I think we need to introduce an `AnonClassSym` or special hash function for anonymous classes, since the default hash includes e.g., `SourceLocation`. That could come back to bite us someday.

File: main/src/ca/uwaterloo/flix/language/phase/Eraser.scala (hunk 1 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Eraser.scala:30]: This class will be deleted.

File: main/src/ca/uwaterloo/flix/language/ast/ReducedAst.scala [hunk 3]
```diff
@@ -21,10 +21,11 @@ import ca.uwaterloo.flix.language.ast.Purity.Pure
 
 object ReducedAst {
 
-  val empty: Root = Root(Map.empty, Map.empty, None, Map.empty)
+  val empty: Root = Root(Map.empty, Map.empty, Set.empty, None, Map.empty)
 
   case class Root(defs: Map[Symbol.DefnSym, Def],
                   enums: Map[Symbol.EnumSym, Enum],
+                  anonClasses: Set[AnonClass],
```
[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/ReducedAst.scala:28]: Agreed. But I am not introducing a set, it already was.

[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/ReducedAst.scala:28]: Could it be a list?

[the remark this answers is not in the document]
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/ast/ReducedAst.scala:28]: I imagine it could be,

**Pull request merged**